### PR TITLE
feat: add after tool call duration

### LINKED ISSUE #3588: [FEATURE] Add duration field to AfterToolCallEvent
### SDK Language

Python

### Problem Statement

Hook authors who want to make decisions based on how long a tool took to execute (e.g. retry on timeout, log slow tools, implement circuit breakers) have no straightforward way to get this information. The SDK already computes tool duration internally for OpenTelemetry metrics in `_stream_with_trace`, but this data is not exposed to the hooks system. A hook author would need to maintain their own stateful timer across BeforeToolCallEvent and AfterToolCallEvent, correlating by `tool_use_id`. This is fragile, especially when retries are involved since the same `tool_use_id` fires multiple before/after pairs.

### Proposed Solution

Add a `duration: float | None` field to `AfterToolCallEvent` (and `BidiAfterToolCallEvent` for parity). The value is the wall-clock seconds elapsed between the `BeforeToolCallEvent` hook completing and the `AfterToolCallEvent` being constructed i.e. the actual tool execution time, excluding hook overhead.

- Timing is captured inside `ToolExecutor._stream`, which is the method that fires the hook, so both `Agent` and `BidiAgent` paths get duration for free.
- The field defaults to `None` for cases where the tool was never actually executed (cancelled by a `BeforeToolCallEvent` hook, or tool not found in registry).
- This is intentionally a single scalar, not a metrics object. Richer observability data (spans, histograms, token counts) remains in the telemetry layer.

### Use Case

- **Timeout-based retry:** A hook retries a tool call when `event.duration` exceeds a threshold, indicating the tool likely hit a network timeout rather than a logic error.
- **Slow tool alerting:** A hook logs a warning or emits a custom metric when any tool exceeds an SLA budget.
- **Adaptive tool selection:** A multi-tool agent uses duration history to prefer faster tool implementations for time-sensitive requests.
- **Circuit breaker patterns:** A hook tracks rolling duration averages and short-circuits tool calls that are consistently degraded.

### Alternatives Solutions

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by seanalbert**
## Description

Adds a `duration: float | None` field to `AfterToolCallEvent` and `BidiAfterToolCallEvent` that reports the wall-clock seconds spent executing a tool.

Hook authors currently have no straightforward way to access tool execution time. The SDK computes duration internally for OpenTelemetry metrics, but this isn't exposed to the hooks system. Without this field, hook authors must maintain stateful timers across `BeforeToolCallEvent`/`AfterToolCallEvent` pairs, correlating by `tool_use_id`. This is fragile when retries are involved.

The timer starts after `BeforeToolCallEvent` completes and stops before `AfterToolCallEvent` is constructed, measuring actual tool execution time and excluding hook overhead. The field is `None` when the tool was never actually executed (cancelled by a hook or not found in the registry).

This enables hook-based patterns like timeout-driven retries, slow tool alerting, adaptive tool selection, and circuit breakers.

## Related Issues

closes #3588

## Documentation PR

No documentation changes required, the field is self-documenting via its docstring and follows the existing pattern of optional fields on hook events.

## Type of Change

New feature

## Testing

How have you tested the change? Verify that the changes do not break functionality or introduce new warnings.

- [x] I ran `hatch run prepare`

Existing tests updated to assert `duration=ANY` where `AfterToolCallEvent` equality is checked

## Checklist

- [x] I have read the CONTRIBUTING document
- [x] I have reviewed and understand every line of code in this PR, including any generated by AI tools, and I can explain why it works
- [x] My change is focused and reasonably small; I have split unrelated work into separate PRs
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [x] I have updated the documentation accordingly
- [x] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

### CODE REVIEW

File: strands-py/src/strands/tools/executors/_executor.py [hunk 1]
```diff
@@ -237,6 +239,7 @@ async def _stream(
                     _interrupt_state=agent._interrupt_state,
                 )
 
+                tool_start_time = time.time()
```
Comment (Unshure) [inline strands-py/src/strands/tools/executors/_executor.py]: This should probably be `time.monotonic()`

File: strands-py/src/strands/storage/local_file_storage.py [hunk 2]
```diff
@@ -250,5 +250,3 @@ async def _walk_sandbox(self, directory: Path, keys: builtins.list[str]) -> None
             else:
                 rel = os.path.relpath(str(full_path), self._base_dir)
                 keys.append(rel.replace(os.sep, "/"))
-
```
Comment (Unshure) [inline strands-py/src/strands/storage/local_file_storage.py:253]: Can you undo these linting changes

**Review by Unshure — COMMENTED** [review]
Generally inclined to this feature, but you should revert the linting changes that were made.

File: strands-py/tests/strands/tools/executors/test_executor.py
Comment (Unshure) [inline strands-py/tests/strands/tools/executors/test_executor.py:1]: Can you add one more test to validate that the duration is none when the tool is interrupted?

**Review by Unshure — COMMENTED** [review]
Code looks good now, can you add one more test, update the pr and commit to follow conventional commits, and rebase to the latest chagnes. Thanks!

**Review by Unshure — APPROVED** [review]
